### DISCUSSION TITLE
max (de)activations per day/week/month?

### DISCUSSION THREAD

**Discussion by nschloe** [keygen-sh/community#135] [Q&A] [answered]
I would like to allow users to activate/deactivate their own machines, but to avoid licenses being passed around indefinitely, I'd like to restrict the number of (de)activations a user can perform _per_ day, week, or month. Any hints?

**Comment by ezekg** [comment]
Currently, this isn't possible. The only way to limit number of activations/deactivations is through license tokens with a max activation/deactivation limits, but in order to "reset" the count e.g. weekly, you would need to generate a new token, since tokens are immutable. There are plans to [add configurable rate-limits to these actions](https://github.com/keygen-sh/keygen-api/issues/442), but that's not a high priority feature at the moment. But contributions are welcome.

**Reply by nschloe** [reply]
Just that. I'd like to allow users to use different machines, e.g., when buying a new computer or using one in a new lab. Without limiting this is in time, one license could service many people if they activate before use, deactivate after. That's why I'd like to have a time-constraint on it.

**Reply by ezekg** [reply]
The best way to prevent that would be to [activate processes](https://keygen.sh/docs/api/processes/) in addition to machines, which will ensure concurrency is enforced. E.g. if a machine is deactivated while the application is running, current processes will be killed and you can exit the application or show a warning, etc. You could, for example, limit each machine to 1 process. This will ensure sharing is impossible.

**Reply by nschloe** [reply]
I didn't catch this. Would that prevent user A to run her program from 9am to 10am, deactivate, forward her license to user B, and he runs it from 11am to 12am? Then ping-pong the license back and forth until the license expires?

**Reply by ezekg** [reply, marked as answer]
It would not prevent that, no. It would only prevent concurrent usage. You would need to implement your own rate limiting logic for now until that previously mentioned issue is tackled.

**Reply by nschloe** [reply]
_Edit:_ Solved by adding to each license an activation token with a limited number of allowed activations. Tokens auto-updated every so often.
